packages/common/src/tools/queue/TaskQueue.ts

```diff
@@ -0,0 +1,97 @@
+import assert from 'assert'
+
+import { Logger } from '../Logger'
+import { Retries } from './Retries'
+import { Job, ShouldRetry } from './types'
+
+interface TaskQueueOpts<T> {
+  workers?: number
+  shouldRetry?: ShouldRetry<T>
+}
+
+export class TaskQueue<T> {
+  private queue: Job<T>[] = []
+  private busyWorkers = 0
+  private workers: number
+  private shouldRetry: ShouldRetry<T>
+
+  constructor(
+    private executeTask: (task: T) => Promise<void>,
+    private logger: Logger,
+    opts?: TaskQueueOpts<T>,
+  ) {
+    this.workers = opts?.workers ?? 1
+    assert(
+      this.workers > 0 && Number.isInteger(this.workers),
+      'workers needs to be a positive integer',
+    )
+    this.shouldRetry = opts?.shouldRetry ?? Retries.always
```

Comment:
I think having exponential backoff as a default retry strategy would be good. Alternatively specifying this parameter shouldn't be optional so that the choice has to be made every time